R/regressionlinear.R

```diff
@@ -1305,39 +1303,31 @@ RegressionLinear <- function(jaspResults, dataset = NULL, options) {
 }
 
 .linregGetPartAndPartialCorrelation <- function(fit, predictors, dataset, options) {
-  predictors <- .linregRemoveFactors(fit, predictors)
-  cors <- vector("list", length(predictors))
-  for (i in seq_along(predictors)) {
+  formula <- formula(fit)
+  R2      <- summary(fit)[["r.squared"]]
 
-    predictor <- predictors[i]
-    if (.linregIsInteraction(predictor)) {
-      newVar                <- .linregMakeCombinedVariableFromInteraction(predictor, dataset)
-      newVarName            <- gsub(":", ".", predictor, fixed = TRUE)
-      dataset[[newVarName]] <- newVar
-    } else {
-      newVarName            <- predictor
-    }
+  cors <- vector("list", length(predictors))
+  names(cors) <- predictors
 
-    dataset <- na.omit(dataset)
-    predictorsToControlFor <- setdiff(predictors, predictor)
-    if (length(predictorsToControlFor) > 0) {
-      formula1 <- .linregGetFormula(newVarName, predictorsToControlFor, includeConstant = TRUE)
-      formula2 <- .linregGetFormula(options$dependent, predictorsToControlFor, includeConstant = TRUE)
+  for(predictor in predictors) {
 
-      cleanedPredictor <- residuals(lm(formula1, data = dataset))
-      cleanedDependent <- residuals(lm(formula2, data = dataset))
+    # drop the term from the formula and refit the model
+    newFormula <- update(formula, as.formula(sprintf(". ~ . - %s", predictor)))
+    newFit     <- update(fit, formula = newFormula)
+    newR2      <- summary(newFit)[["r.squared"]]
 
-      partCor     <- cor(cleanedPredictor, dataset[[options$dependent]])
-      partialCor  <- cor(cleanedPredictor, cleanedDependent)
-    } else { # if there are no variables to control for, return regular correlation
-      partCor <- partialCor <- cor(dataset[[options$dependent]], dataset[[newVarName]])
-    }
+    sr2 <- R2 - newR2      # squared semi-partial (part) correlation
+    pr2 <- sr2 / (1-newR2) # squared partial correlation
 
-    cors[[i]] <- list(
-      name    = .unvf(predictor),
-      part    = partCor,
-      partial = partialCor)
+    # determine the sign of the coefficient
+    sign <- sign(coefficients(fit)[predictor])
+    if(is.na(sign)) sign <- 1 # for categorical predictors
 
+    cors[[predictor]] <- list(
+      name    = predictor,
```

Comment:
I think this should be
```suggestion
      name    = jaspBase::gsubInteractionSymbol(predictor),
```
because now I'm seeing 
![image](https://user-images.githubusercontent.com/21319932/156890937-abc83b92-c0ed-416f-84ad-62fe43abe65a.png)

rather than

![image](https://user-images.githubusercontent.com/21319932/156890947-2f85d398-d65b-4268-a560-e15ab416becd.png)

This wasn't introduced by you though.